sdk/sync/watch.go

```diff
@@ -61,43 +65,25 @@ func (w *Watcher) Subscribe(subtree *Subtree, ch interface{}) (func() error, err
 		return nil, err
 	}
 
-	w.lk.Lock()
-
-	// Make sure we have a subtree mapping.
-	if _, ok := w.subtrees[subtree]; !ok {
-		w.subtrees[subtree] = make(map[*subscription]struct{})
-	}
+	ctx, cancel := context.WithCancel(ctx)
 
 	root := w.root + ":" + subtree.GroupKey
 	sub := &subscription{
 		w:       w,
 		subtree: subtree,
-		client:  w.client,
+		client:  w.client.WithContext(ctx),
 		key:     root,
-		connCh:  make(chan int64, 1),
-		closeCh: make(chan struct{}),
-		doneCh:  make(chan struct{}),
 		outCh:   chV,
 	}
 
-	w.subtrees[subtree][sub] = struct{}{}
-	w.lk.Unlock()
-
 	// Start the subscription.
-	go sub.process()
-
-	cancelFn := func() error {
-		w.lk.Lock()
-		defer w.lk.Unlock()
-
-		delete(w.subtrees[subtree], sub)
-		if len(w.subtrees[subtree]) == 0 {
-			delete(w.subtrees, subtree)
-		}
+	w.subs.Add(1)
+	go func() {
+		defer w.subs.Done()
+		sub.process()
+	}()
```

Comment:
The subscriptions will automatically shutdown when the context is canceled. No need to track them.